### PR TITLE
os/bluestore: kv_max -> kv_min

### PR DESCRIPTION
**Pull request by liewegas**
- change config option to a minimum instead of a maximum, and assume all memory goes to rocksdb below the min
- change the meta vs kv balance to 50/50 above the min until we have better information.

Notably we still spend nothing on data.  I wonder if we should give it like 1% or something by default so that there is *some* caching of the hottest objects.

### CODE REVIEW

**Comment by liewegas** [comment]
retest this please

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-02-25_02:17:51-rados-wip-kefu-testing-2018-02-25-0801-distro-basic-mira/

File: src/common/options.cc [hunk 1]
```diff
@@ -3735,19 +3735,24 @@ std::vector<Option> get_global_options() {
...
-    .set_default(.99)
+    .set_default(.01)
+    .add_see_also("bluestore_cache_size")
     .set_description("Ratio of bluestore cache to devote to kv database (rocksdb)"),
 
-    Option("bluestore_cache_kv_max", Option::TYPE_INT, Option::LEVEL_ADVANCED)
+    Option("bluestore_cache_kv_min", Option::TYPE_INT, Option::LEVEL_ADVANCED)
     .set_default(512_M)
-    .set_description("Max memory (bytes) to devote to kv database (rocksdb)")
+    .set_description("Minimm memory (bytes) of bluestore_cache_size to devote to kv database (rocksdb)")
```
Comment (ifed01) [inline src/common/options.cc]: typo

File: src/common/options.cc [hunk 2]
```diff
@@ -3735,19 +3735,24 @@ std::vector<Option> get_global_options() {
...
+    .add_see_also("bluestore_cache_size")
     .set_long_description("A negative value means using bluestore_cache_meta_ratio "
       "and bluestore_cache_kv_ratio instead of calculating these ratios using "
-      "bluestore_cache_size_* and bluestore_cache_kv_max."),
+      "bluestore_cache_size_* and bluestore_cache_kv_max.  If "
```
Comment (ifed01) [inline src/common/options.cc]: max -> min

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -3904,6 +3886,26 @@ int BlueStore::_set_cache_sizes()
...
+
+  // if cache_kv_min is negative, disable it
+  if (cache_size > 0 && cache_kv_min >= 0) {
+    cache_kv_min_ratio = std::max((double)cache_kv_min / (double)cache_size,
+				  (double)1.0);
+    if (cache_kv_min_ratio > cache_kv_ratio) {
+      dout(1) << __func__ << " kv_min_ratio " << cache_kv_min_ratio
+            << " > kv_ratio " << cache_kv_ratio
+            << dendl;
+      cache_meta_ratio = cache_meta_ratio + cache_kv_ratio - cache_kv_min_ratio;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: IMO that's not 100% valid formula.  E.g. for cache_meta_ratio = 0.3  and cache_kv_ratio = 0.2 and cache_kv_min_ratio = 1.0 it results in cache_meta_ratio = -0.5 and subsequent cache_data_ratio = 0.5

Comment (ifed01) [reply src/os/bluestore/BlueStore.cc]: BTW this formula also implies that additional space given to kv cache is taken from meta one. Preserving data cache ratio as is (to some degree), e.g. for the case cache_meta_ratio = 0.5 and cache_kv_ratio = 0.3 and cache_kv_min_ratio = 0.5 we'll get cache_meta_ratio decrease while cache_data_ratio is preserved at 0.2 .
Not sure that's the best policy.
Maybe something like 
cache_meta_ratio = min(cache_meta_ratio, 1.0 - cache_kv_min_ratio)
?

File: src/common/options.cc (hunk 2 above)
Comment (ifed01) [reply src/common/options.cc]: And do we need that  mention about negative value? Wouldn't setting 'bluestore_cache_kv_min' to zero cause the same behavior ( due to 'if cache_kv_min_ratio > cache_kv_ratio' statement below?

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -3904,6 +3886,25 @@ int BlueStore::_set_cache_sizes()
 	 << dendl;
     return -EINVAL;
   }
+
+  double cache_kv_min = cct->_conf->bluestore_cache_kv_min;
+  double cache_kv_min_ratio = 0;
+
+  // if cache_kv_min is negative, disable it
+  if (cache_size > 0 && cache_kv_min >= 0) {
+    cache_kv_min_ratio = std::max((double)cache_kv_min / (double)cache_size,
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: min not max?

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -3904,6 +3886,26 @@ int BlueStore::_set_cache_sizes()
...
+
+  // if cache_kv_min is negative, disable it
+  if (cache_size > 0 && cache_kv_min >= 0) {
+    cache_kv_min_ratio = std::min((double)cache_kv_min / (double)cache_size,
+				  (double)1.0);
+    if (cache_kv_min_ratio > cache_kv_ratio) {
+      dout(1) << __func__ << " kv_min_ratio " << cache_kv_min_ratio
+            << " > kv_ratio " << cache_kv_ratio << dendl;
+      cache_kv_ratio = cache_kv_min_ratio;
+      cache_meta_ratio = std::min(cache_meta_ratio,
```
Comment (tchaikov) [inline src/os/bluestore/BlueStore.cc]:
```
/build/ceph-13.0.1-2341-g20782c9/src/os/bluestore/BlueStore.cc: In member function 'int BlueStore::_set_cache_sizes()':
/build/ceph-13.0.1-2341-g20782c9/src/os/bluestore/BlueStore.cc:3921:35: error: no matching function for call to 'min(float&, double)'
       (double)1.0 - cache_kv_ratio);
                                   ^
```

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-03-05_17:31:09-rados-wip-kefu-testing-2018-03-05-2238-distro-basic-smithi/

**Pull request merged**